simulariumio/converter.py

```diff
@@ -0,0 +1,180 @@
+
+log = logging.getLogger(__name__)
+
+###############################################################################
+
+SUPPORTED_TRAJECTORY_READERS = {
+    "custom": CustomTrajectoryReader,
+}
+
+SUPPORTED_PLOT_READERS = {
+    "scatter": ScatterPlotReader,
+    "histogram": HistogramPlotReader,
+}
+
+###############################################################################
+
+
+class Converter:
+    _data: Dict[str, Any] = {}
+
+    def __init__(self, data: Dict[str, Any] = {}, source_engine: str = "custom"):
+        """
+        This object reads simulation trajectory outputs from various engines 
+        (as well as custom data) and plot data and writes them 
+        in the JSON format used by the Simularium viewer.
+
+        Parameters
+        ----------
+        data: Dict[str, Any]
+            Loaded data or path to data from a simulation engine. 
+            Fields for each engine:
+
+                custom: 
+                    box_size: np.ndarray (shape = [3])
+                        A numpy ndarray containing the XYZ dimensions 
+                        of the simulation bounding volume
+                    times: np.ndarray (shape = [timesteps])
+                        A numpy ndarray containing the elapsed simulated time 
+                        at each timestep
+                    n_agents: np.ndarray (shape = [timesteps])
+                        A numpy ndarray containing the number of agents 
+                        that exist at each timestep
+                    viz_types : np.ndarray (shape = [timesteps, agents])
+                        A numpy ndarray containing the viz type 
+                        for each agent at each timestep. Current options:
+                            1000 : default,
+                            1001 : fiber (which will require subpoints)
+                    positions : np.ndarray (shape = [timesteps, agents, 3])
+                        A numpy ndarray containing the XYZ position 
+                        for each agent at each timestep
+                    types: List[List[str]] (list of shape [timesteps, agents])
+                        A list containing timesteps, for each a list of 
+                        the string name for the type of each agent
+                    radii : np.ndarray (shape = [timesteps, agents])
+                        A numpy ndarray containing the radius 
+                        for each agent at each timestep
+                    subpoints: np.ndarray 
+                    (shape = [timesteps, agents, subpoints, 3]) (optional) 
+                        A numpy ndarray containing a list of subpoint position data 
+                        for each agent at each timestep. These values are 
+                        currently only used for fiber agents.
+                    plots : Dict[str, Any] (optional) 
+                        An object containing plot data already 
+                        in Simularium format
```

Comment:
It may be better to make an entire documentation page about this format rather than including this in the docstring as it would allow more rich formatting. And then in the docstring simply say "go look at docs for format"

Reply:
Would love @toloudis opinion though.

Reply:
I guess that changing the "Any" to some kind of python data structure definition would be my preferred thing.  But I don't know how expressive python's types can get, like can you specify "optional"?  And yes, this spec is the crux of the whole thing so it really needs to be easily findable in documentation.  But in my opinion, I think it would be easier to keep up-to-date with changes if it stays here near where it is used in the code.  

Also, it is not clear what the `str` key represents for the passed-in data dict.  Also, what does "loaded data or path to data" mean?

Reply:
I think keeping the documentation near the code is a great idea FWIW.

The `str` is the key, how should I specify that? I'm trying to list them here...

by "loaded data or path to data" I mean that this object passed in might have numpy arrays of data in it, or it might have a filepath to where to load the data. Basically, for custom data like this, a user will have to shape their data into numpy arrays, but for specific simulation engines we support (which I'll be adding shortly) we'll know their output format so can load and shape the data for the user. Should I say this differently? Or just not say it since the keys indicate it?